crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs

```diff
@@ -362,111 +348,67 @@ impl RetriableRpcClient {
-                );
-                return Err(error);
+        match self.download_handler.download_checkpoint_decision() {
+            CheckpointDownloadDecision::AttemptPrimary => {
+                self.handle_primary_attempt(sequence_number).await
+            }
+            CheckpointDownloadDecision::TryFallback(_) => {
+                self.handle_fallback_attempt(sequence_number).await
+            }
+            CheckpointDownloadDecision::UsePrimaryError(_) => {
+                unreachable!("Invalid initial decision")
             }
         }
-
-        let fallback_client = self
-            .fallback_client
-            .as_ref()
-            .expect("fallback client must set");
-        self.get_checkpoint_from_fallback(sequence_number, fallback_client)
-            .await
     }
 
-    /// Gets the full checkpoint data for the given sequence number from the fallback archival.
-    async fn get_checkpoint_from_fallback(
+    /// Handles the primary RPC attempt and potential fallback
+    async fn handle_primary_attempt(
         &self,
         sequence_number: u64,
-        fallback_client: &FallbackClient,
     ) -> Result<CheckpointData, RetriableClientError> {
-        // RPC node failure is sustained, so we skip it for `skip_rpc_for_checkpoint_duration`.
-        // Also, at this point, the fallback client must be set.
-        if fallback_client.is_failure_window_exceeded(
-            self.last_success.load(Ordering::Relaxed),
-            self.num_failures.load(Ordering::Relaxed),
-        ) {
-            self.skip_rpc_for_checkpoint_until.store(
-                Instant::now()
-                    + self
-                        .fallback_client
-                        .as_ref()
-                        .expect("fallback client must set")
-                        .skip_rpc_for_checkpoint_duration(),
-                Ordering::Relaxed,
-            );
-        }
-
-        let fallback_start_time = Instant::now();
-        let result = self
-            .call_get_full_checkpoint_from_fallback_with_retries(fallback_client, sequence_number)
-            .await;
-
-        if let Some(metrics) = self.metrics.as_ref() {
-            metrics.record_fallback_metrics(
-                "get_full_checkpoint",
-                &result,
-                fallback_start_time.elapsed(),
-            )
+        match self.get_full_checkpoint_from_primary(sequence_number).await {
+            Ok(checkpoint) => {
+                self.download_handler
+                    .record_primary_success(sequence_number);
+                Ok(checkpoint)
+            }
+            Err(primary_error) => {
+                match self
+                    .download_handler
+                    .handle_primary_checkpoint_failure(sequence_number, primary_error)
+                {
+                    CheckpointDownloadDecision::TryFallback(Some(_)) => {
+                        self.handle_fallback_attempt(sequence_number).await
```

Comment:
Similarly here I wonder if it would be helpful to log the error? If we decide to always ignore the error, then we could remove it from the `CheckpointDownloadDecision::TryFallback` enum value.